File: backend/app/schemas/workflow.py
"""
Workflow Management Schemas

Pydantic schemas for workflow management API.
Based on docs/design/workflow-api-design.md
"""

from datetime import datetime
from typing import Optional, List, Dict, Any, Literal
from pydantic import BaseModel, Field
from enum import Enum


# ============================================================
# Enums
# ============================================================

class WorkflowStatus(str, Enum):
    """工作流状态"""
    PENDING = "pending"
    RUNNING = "running"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"
    TERMINATED = "terminated"


class StepStatus(str, Enum):
    """步骤状态"""
    PENDING = "pending"
    READY = "ready"
    ASSIGNED = "assigned"
    RUNNING = "running"
    AWAITING_REVIEW = "awaiting_review"
    APPROVED = "approved"
    REJECTED = "rejected"
    RETRYING = "retrying"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    SKIPPED = "skipped"


class ReviewAction(str, Enum):
    """审核动作"""
    APPROVE = "approve"
    REJECT = "reject"
    REQUEST_CHANGES = "request_changes"


class AgentStatus(str, Enum):
    """Agent状态"""
    ONLINE = "online"
    DEGRADED = "degraded"
    OFFLINE = "offline"


class TemplateStatus(str, Enum):
    """模板状态"""
    DRAFT = "draft"
    PUBLISHED = "published"
    ARCHIVED = "archived"


# ============================================================
# Common Schemas
# ============================================================

class PaginationParams(BaseModel):
    """分页参数"""
    page: int = Field(1, ge=1, description="页码")
    page_size: int = Field(20, ge=1, le=100, description="每页数量")


class PaginatedResponse(BaseModel):
    """分页响应"""
    total: int = Field(..., description="总数")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页数量")
    total_pages: int = Field(..., description="总页数")


class ErrorResponse(BaseModel):
    """错误响应"""
    error: Dict[str, Any] = Field(..., description="错误信息")


# ============================================================
# Workflow Template Schemas
# ============================================================

class StepNode(BaseModel):
    """DAG中的步骤节点"""
    id: str = Field(..., description="步骤 ID")
    name: str = Field(..., description="步骤名称")
    agent: Optional[str] = Field(None, description="Agent 名称或 ID")
    capabilities: Optional[List[str]] = Field(None, description="Agent 能力标签")
    estimated_duration: Optional[int] = Field(None, description="预估时长（分钟）")
    output: Optional[str] = Field(None, description="输出文件路径")
    validation: Optional[List[str]] = Field(None, description="验证规则")
    human_review: bool = Field(False, description="是否需要人工审核")
    depends_on: List[str] = Field(default_factory=list, description="依赖的步骤 ID")
    input_schema: Optional[Dict[str, Any]] = Field(None, description="输入参数 schema")
    output_schema: Optional[Dict[str, Any]] = Field(None, description="输出参数 schema")


class EdgeNode(BaseModel):
    """DAG中的边"""
    source: str = Field(..., description="源步骤 ID")
    target: str = Field(..., description="目标步骤 ID")


class DAGDefinition(BaseModel):
    """DAG 定义"""
    steps: List[StepNode] = Field(..., description="步骤列表")
    edges: List[EdgeNode] = Field(..., description="边列表")


class WorkflowConfig(BaseModel):
    """工作流配置"""
    single_step_timeout: Optional[int] = Field(1800, description="单步超时（秒）")
    workflow_timeout: Optional[int] = Field(86400, description="工作流超时（秒）")
    max_retries: Optional[int] = Field(3, description="最大重试次数")
    failure_strategy: Optional[str] = Field("escalate", description="失败策略")


class WorkflowTemplateCreate(BaseModel):
    """创建工作流模板请求"""
    name: str = Field(..., description="模板名称")
    description: Optional[str] = Field(None, description="模板描述")
    dag: DAGDefinition = Field(..., description="DAG 定义")
    config: Optional[WorkflowConfig] = Field(None, description="工作流配置")
    tags: Optional[List[str]] = Field(None, description="标签列表")


class WorkflowTemplateUpdate(BaseModel):
    """更新工作流模板请求"""
    name: Optional[str] = Field(None, description="模板名称")
    description: Optional[str] = Field(None, description="模板描述")
    dag: Optional[DAGDefinition] = Field(None, description="DAG 定义")
    config: Optional[WorkflowConfig] = Field(None, description="工作流配置")
    tags: Optional[List[str]] = Field(None, description="标签列表")


class WorkflowTemplateResponse(BaseModel):
    """工作流模板响应"""
    id: str
    name: str
    description: Optional[str]
    version: str
    status: TemplateStatus
    dag: DAGDefinition
    config: WorkflowConfig
    created_at: str
    created_by: str
    updated_at: str
    published_at: Optional[str]
    usage_count: int
    tags: Optional[List[str]]
    steps: Optional[List[StepNode]] = None

    class Config:
        from_attributes = True


class WorkflowTemplateListResponse(PaginatedResponse):
    """工作流模板列表响应"""
    data: List[WorkflowTemplateResponse]


# ============================================================
# Workflow Instance Schemas
# ============================================================

class CurrentStepInfo(BaseModel):
    """当前步骤信息"""
    step_id: str
    name: str
    status: StepStatus
    agent_name: Optional[str]
    progress: int


class WorkflowInstanceCreate(BaseModel):
    """创建工作流实例请求"""
    template_id: str = Field(..., description="模板 ID")
    input: Dict[str, Any] = Field(default_factory=dict, description="输入参数")
    execution_mode: Optional[str] = Field("standard", description="执行模式")


class WorkflowInstanceResponse(BaseModel):
    """工作流实例响应"""
    id: str
    template_id: str
    template_name: Optional[str] = None
    template_version: str
    status: WorkflowStatus
    input: Dict[str, Any]
    output: Optional[Dict[str, Any]]
    progress: int
    estimated_remaining: Optional[int]
    created_at: str
    created_by: str
    started_at: Optional[str]
    completed_at: Optional[str]
    duration: Optional[int]
    error_message: Optional[str]
    termination_reason: Optional[str]
    current_step: Optional[CurrentStepInfo] = None
    steps: Optional[List["StepExecutionResponse"]] = None

    class Config:
        from_attributes = True


class WorkflowInstanceListResponse(PaginatedResponse):
    """工作流实例列表响应"""
    data: List[WorkflowInstanceResponse]


class TerminateWorkflowRequest(BaseModel):
    """终止工作流请求"""
    reason: str = Field(..., description="终止原因")


# ============================================================
# Step Execution Schemas
# ============================================================

class StepExecutionResponse(BaseModel):
    """步骤执行响应"""
    id: str
    workflow_instance_id: str
    step_id: str
    name: str
    status: StepStatus
    agent_id: Optional[str]
    agent_name: Optional[str]
    input: Optional[Dict[str, Any]]
    output: Optional[Dict[str, Any]]
    progress: int
    progress_message: Optional[str]
    started_at: Optional[str]
    completed_at: Optional[str]
    duration: Optional[int]
    retry_count: int
    max_retries: int
    error_message: Optional[str]
    force_completed: bool = False
    human_review: bool = False
    review: Optional["ReviewResponse"] = None
    artifacts: Optional[List["ArtifactResponse"]] = None

    class Config:
        from_attributes = True


class StepExecutionListResponse(PaginatedResponse):
    """步骤执行列表响应"""
    data: List[StepExecutionResponse]


class RetryStepRequest(BaseModel):
    """重试步骤请求"""
    pass


class SkipStepRequest(BaseModel):
    """跳过步骤请求"""
    reason: str = Field(..., description="跳过原因")


class ForceCompleteStepRequest(BaseModel):
    """强制完成步骤请求"""
    reason: str = Field(..., description="强制完成原因")
    output: Optional[Dict[str, Any]] = Field(None, description="输出数据")


class ReassignAgentRequest(BaseModel):
    """重新分配 Agent 请求"""
    agent_id: str = Field(..., description="Agent ID")


class ProgressUpdateRequest(BaseModel):
    """Agent 上报进度请求"""
    progress: int = Field(..., ge=0, le=100, description="进度百分比")
    message: Optional[str] = Field(None, description="进度消息")
    estimated_remaining: Optional[int] = Field(None, description="预估剩余时间（秒）")


class StepInputResponse(BaseModel):
    """步骤输入响应"""
    input: Dict[str, Any]
    dependencies: List[Dict[str, Any]]


class StepOutputResponse(BaseModel):
    """步骤输出响应"""
    output: Dict[str, Any]
    artifacts: List["ArtifactResponse"]


# ============================================================
# Review Schemas
# ============================================================

class ReviewResponse(BaseModel):
    """审核记录响应"""
    id: str
    workflow_instance_id: str
    workflow_name: Optional[str] = None
    step_execution_id: str
    step_name: Optional[str] = None
    reviewer_id: str
    reviewer_name: Optional[str]
    action: Optional[ReviewAction]
    comment: Optional[str]
    created_at: str
    updated_at: str
    timeout_at: Optional[str]
    timeout_action: str
    remaining_time: Optional[int]
    review_round: int

    class Config:
        from_attributes = True


class ReviewDetailResponse(ReviewResponse):
    """审核详情响应"""
    outputs: Optional[Dict[str, Any]] = None
    history: Optional[List[Dict[str, Any]]] = None


class ReviewListResponse(PaginatedResponse):
    """审核列表响应"""
    data: List[ReviewResponse]


class ApproveReviewRequest(BaseModel):
    """通过审核请求"""
    comment: Optional[str] = Field(None, description="审核意见")


class RejectReviewRequest(BaseModel):
    """拒绝审核请求"""
    comment: str = Field(..., description="拒绝原因")


class RequestChangesReviewRequest(BaseModel):
    """要求修改请求"""
    comment: str = Field(..., description="修改要求")


class ReviewStatsResponse(BaseModel):
    """审核统计响应"""
    total_pending: int
    total_completed_today: int
    timeout_warnings: int
    avg_review_time_seconds: int
    by_action: Dict[str, int]


# ============================================================
# Agent Schemas
# ============================================================

class AgentResponse(BaseModel):
    """Agent 响应"""
    id: str
    name: str
    display_name: Optional[str]
    capabilities: List[str]
    status: AgentStatus
    current_task_id: Optional[str]
    current_workflow_instance_id: Optional[str]
    current_step_execution_id: Optional[str]
    last_heartbeat: Optional[str]
    config: Optional[Dict[str, Any]]
    metadata: Optional[Dict[str, Any]]
    statistics: Optional[Dict[str, Any]] = None
    created_at: str
    updated_at: str

    class Config:
        from_attributes = True


class AgentListResponse(PaginatedResponse):
    """Agent 列表响应"""
    data: List[AgentResponse]


class AgentLoadStatsResponse(BaseModel):
    """Agent 负载统计响应"""
    agent_id: str
    agent_name: str
    status: AgentStatus
    current_tasks: int
    running_tasks: int
    avg_task_duration: int


class AgentLoadStatsListResponse(BaseModel):
    """Agent 负载统计列表响应"""
    data: List[AgentLoadStatsResponse]
    total_agents: int
    online_agents: int
    offline_agents: int


class BatchCleanupRequest(BaseModel):
    """批量清理请求"""
    agent_ids: List[str] = Field(..., description="Agent ID 列表")


# ============================================================
# Artifact Schemas
# ============================================================

class ArtifactResponse(BaseModel):
    """产出物响应"""
    id: str
    workflow_instance_id: str
    step_execution_id: Optional[str]
    artifact_type: str
    name: str
    description: Optional[str]
    storage_kind: str
    storage_path: Optional[str]
    size_bytes: Optional[int]
    checksum: Optional[str]
    created_at: str

    class Config:
        from_attributes = True


class ArtifactListResponse(PaginatedResponse):
    """产出物列表响应"""
    data: List[ArtifactResponse]


# ============================================================
# Log Schemas
# ============================================================

class LogResponse(BaseModel):
    """日志响应"""
    id: str
    step_execution_id: Optional[str]
    timestamp: str
    level: str
    message: str
    metadata: Optional[Dict[str, Any]]

    class Config:
        from_attributes = True


class LogListResponse(PaginatedResponse):
    """日志列表响应"""
    data: List[LogResponse]


# ============================================================
# Event Schemas
# ============================================================

class EventResponse(BaseModel):
    """事件响应"""
    id: str
    event_type: str
    timestamp: str
    actor_type: Optional[str]
    actor_id: Optional[str]
    event_data: Optional[Dict[str, Any]]

    class Config:
        from_attributes = True


class EventListResponse(PaginatedResponse):
    """事件列表响应"""
    data: List[EventResponse]


# ============================================================
# Statistics Schemas
# ============================================================

class WorkflowStatsResponse(BaseModel):
    """工作流统计响应"""
    total: int
    by_status: Dict[str, int]
    success_rate: float
    avg_duration_seconds: int
    by_template: List[Dict[str, Any]]
    time_series: List[Dict[str, Any]]


class AgentStatsResponse(BaseModel):
    """Agent 统计响应"""
    total_agents: int
    online_agents: int
    offline_agents: int
    by_agent: List[Dict[str, Any]]
    total_tasks: int
    total_tokens: int
    total_cost_usd: float


class TaskStatsResponse(BaseModel):
    """任务统计响应"""
    total: int
    by_status: Dict[str, int]
    by_category: Dict[str, int]
    avg_duration_seconds: int
    time_series: List[Dict[str, Any]]


class HealthCheckResponse(BaseModel):
    """健康检查响应"""
    status: str
    timestamp: str
    components: Dict[str, Any]
    version: str


# ============================================================
# WebSocket Schemas
# ============================================================

class WebSocketSubscribeMessage(BaseModel):
    """WebSocket 订阅消息"""
    action: Literal["subscribe", "unsubscribe"]
    channel: str


class WorkflowEventMessage(BaseModel):
    """工作流事件消息"""
    event: str
    channel: str
    timestamp: str
    data: Dict[str, Any]


# ============================================================
# Callback Schemas (Internal)
# ============================================================

class StepCallbackRequest(BaseModel):
    """步骤完成回调请求"""
    status: StepStatus
    output: Optional[Dict[str, Any]] = None
    artifacts: Optional[List[Dict[str, Any]]] = None
    metrics: Optional[Dict[str, Any]] = None


class StepProgressCallbackRequest(BaseModel):
    """步骤进度回调请求"""
    progress: int = Field(..., ge=0, le=100)
    message: Optional[str] = None
    estimated_remaining: Optional[int] = None


# ============================================================
# Template Version Schemas
# ============================================================

class TemplateVersionResponse(BaseModel):
    """模板版本响应"""
    version: str
    change_summary: Optional[str]
    created_at: str
    created_by: str


class TemplateVersionListResponse(PaginatedResponse):
    """模板版本列表响应"""
    data: List[TemplateVersionResponse]


class RollbackRequest(BaseModel):
    """回滚请求"""
    version: str = Field(..., description="目标版本")


# ============================================================
# Duplicate Template Schemas
# ============================================================

class DuplicateTemplateRequest(BaseModel):
    """复制模板请求"""
    name: str = Field(..., description="新模板名称")
    description: Optional[str] = Field(None, description="新模板描述")


# ============================================================
# Export Schemas
# ============================================================

class ExportOptions(BaseModel):
    """导出选项"""
    format: Optional[str] = Field("json", description="导出格式")

File: frontend/src/types/workflow.ts
/**
 * 工作流管理系统 - TypeScript 类型定义
 * 基于 API 设计文档 v1.0
 */

// ==================== 枚举定义 ====================

/**
 * 工作流状态
 */
export enum WorkflowStatus {
  PENDING = 'pending',
  RUNNING = 'running',
  PAUSED = 'paused',
  COMPLETED = 'completed',
  FAILED = 'failed',
  TERMINATED = 'terminated'
}

/**
 * 步骤状态
 */
export enum StepStatus {
  PENDING = 'pending',
  READY = 'ready',
  ASSIGNED = 'assigned',
  RUNNING = 'running',
  AWAITING_REVIEW = 'awaiting_review',
  APPROVED = 'approved',
  REJECTED = 'rejected',
  RETRYING = 'retrying',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
  SKIPPED = 'skipped'
}

/**
 * 审核动作
 */
export enum ReviewAction {
  APPROVE = 'approve',
  REJECT = 'reject',
  REQUEST_CHANGES = 'request_changes'
}

/**
 * Agent 状态
 */
export enum AgentStatus {
  ONLINE = 'online',
  DEGRADED = 'degraded',
  OFFLINE = 'offline'
}

/**
 * 模板状态
 */
export enum TemplateStatus {
  DRAFT = 'draft',
  PUBLISHED = 'published',
  ARCHIVED = 'archived'
}

/**
 * 失败策略
 */
export enum FailureStrategy {
  RETRY = 'retry',
  SKIP = 'skip',
  ESCALATE = 'escalate',
  TERMINATE = 'terminate'
}

/**
 * 执行模式
 */
export enum ExecutionMode {
  STANDARD = 'standard',
  FAST_TRACK = 'fast_track',
  DEBUG = 'debug'
}

// ==================== 通用类型 ====================

/**
 * 分页请求参数
 */
export interface PaginationParams {
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

/**
 * 分页响应
 */
export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

/**
 * 错误响应
 */
export interface ErrorResponse {
  error: {
    code: string;
    message: string;
    details?: Record<string, any>;
  };
}

/**
 * 成功响应
 */
export interface SuccessResponse {
  success: boolean;
  message?: string;
}

// ==================== DAG 相关类型 ====================

/**
 * DAG 步骤定义
 */
export interface DAGStep {
  id: string;
  name: string;
  agent: string;
  capabilities?: string[];
  estimated_duration?: number;
  output?: string;
  validation?: string[];
  human_review?: boolean;
  depends_on: string[];
}

/**
 * DAG 边定义
 */
export interface DAGEdge {
  source: string;
  target: string;
}

/**
 * DAG 结构
 */
export interface DAG {
  steps: DAGStep[];
  edges: DAGEdge[];
}

/**
 * 工作流配置
 */
export interface WorkflowConfig {
  single_step_timeout?: number;
  workflow_timeout?: number;
  max_retries?: number;
  failure_strategy?: FailureStrategy;
}

// ==================== 工作流模板相关类型 ====================

/**
 * 工作流模板（列表项）
 */
export interface WorkflowTemplateListItem {
  id: string;
  name: string;
  description?: string;
  version: string;
  status: TemplateStatus;
  usage_count: number;
  tags?: string[];
  created_at: string;
  created_by: string;
  updated_at: string;
  published_at?: string;
  step_count: number;
}

/**
 * 工作流模板（详情）
 */
export interface WorkflowTemplate extends WorkflowTemplateListItem {
  dag: DAG;
  config: WorkflowConfig;
  steps: Array<{
    id: string;
    name: string;
    agent: string;
    estimated_duration?: number;
    human_review?: boolean;
  }>;
}

/**
 * 创建模板请求
 */
export interface CreateTemplateRequest {
  name: string;
  description?: string;
  dag: DAG;
  config?: WorkflowConfig;
  tags?: string[];
}

/**
 * 更新模板请求
 */
export interface UpdateTemplateRequest extends CreateTemplateRequest {}

/**
 * 模板列表查询参数
 */
export interface TemplateListParams extends PaginationParams {
  status?: TemplateStatus;
  search?: string;
  tags?: string;
  created_by?: string;
}

/**
 * 模板版本
 */
export interface TemplateVersion {
  version: string;
  change_summary: string;
  created_at: string;
  created_by: string;
}

// ==================== 工作流实例相关类型 ====================

/**
 * 当前步骤信息
 */
export interface CurrentStep {
  step_id: string;
  name: string;
  status: StepStatus;
  agent_name: string;
  progress?: number;
}

/**
 * 工作流实例（列表项）
 */
export interface WorkflowInstanceListItem {
  id: string;
  template_id: string;
  template_name: string;
  template_version: string;
  status: WorkflowStatus;
  input: Record<string, any>;
  output?: Record<string, any>;
  progress: number;
  estimated_remaining?: number;
  created_at: string;
  created_by: string;
  started_at?: string;
  completed_at?: string;
  duration?: number;
  current_step?: CurrentStep;
}

/**
 * 步骤执行记录
 */
export interface StepExecution {
  id: string;
  step_id: string;
  name: string;
  status: StepStatus;
  agent_id?: string;
  agent_name?: string;
  input?: Record<string, any>;
  output?: Record<string, any>;
  progress?: number;
  progress_message?: string;
  started_at?: string;
  completed_at?: string;
  duration?: number;
  retry_count: number;
  max_retries?: number;
  error_message?: string;
  force_completed?: boolean;
  human_review?: boolean;
}

/**
 * 工作流实例（详情）
 */
export interface WorkflowInstance extends WorkflowInstanceListItem {
  error_message?: string;
  termination_reason?: string;
  steps: StepExecution[];
}

/**
 * 启动工作流请求
 */
export interface StartWorkflowRequest {
  template_id: string;
  input: Record<string, any>;
  execution_mode?: ExecutionMode;
}

/**
 * 终止工作流请求
 */
export interface TerminateWorkflowRequest {
  reason: string;
}

/**
 * 工作流实例列表查询参数
 */
export interface WorkflowListParams extends PaginationParams {
  status?: WorkflowStatus;
  template_id?: string;
  created_by?: string;
  start_date?: string;
  end_date?: string;
}

/**
 * 日志条目
 */
export interface LogEntry {
  id: string;
  step_execution_id?: string;
  timestamp: string;
  level: 'INFO' | 'WARN' | 'ERROR' | 'DEBUG';
  message: string;
  metadata?: Record<string, any>;
}

/**
 * 日志查询参数
 */
export interface LogQueryParams {
  level?: 'INFO' | 'WARN' | 'ERROR' | 'DEBUG';
  step_id?: string;
  start_time?: string;
  end_time?: string;
  search?: string;
  limit?: number;
}

/**
 * 事件时间线条目
 */
export interface TimelineEvent {
  id: string;
  event_type: string;
  timestamp: string;
  actor_type: 'user' | 'agent' | 'system';
  actor_id: string;
  event_data: Record<string, any>;
}

// ==================== 步骤执行相关类型 ====================

/**
 * 审核信息
 */
export interface ReviewInfo {
  id: string;
  reviewer_id: string;
  reviewer_name: string;
  action?: ReviewAction;
  timeout_at: string;
  remaining_time: number;
}

/**
 * 产出物信息
 */
export interface Artifact {
  id: string;
  name: string;
  artifact_type: string;
  size_bytes: number;
  storage_path?: string;
  download_url?: string;
  preview_url?: string;
}

/**
 * 步骤详情（扩展）
 */
export interface StepExecutionDetail extends StepExecution {
  workflow_instance_id: string;
  review?: ReviewInfo;
  artifacts?: Artifact[];
}

/**
 * 重试步骤请求
 */
export interface RetryStepRequest {
  // 无需参数
}

/**
 * 跳过步骤请求
 */
export interface SkipStepRequest {
  reason: string;
}

/**
 * 强制完成步骤请求
 */
export interface ForceCompleteStepRequest {
  reason: string;
  output?: Record<string, any>;
}

/**
 * 重新分配 Agent 请求
 */
export interface ReassignAgentRequest {
  agent_id: string;
}

/**
 * 上报进度请求
 */
export interface ReportProgressRequest {
  progress: number;
  message?: string;
  estimated_remaining?: number;
}

/**
 * 步骤输出
 */
export interface StepOutput {
  output: Record<string, any>;
  artifacts: Artifact[];
}

/**
 * 步骤输入
 */
export interface StepInput {
  input: Record<string, any>;
  dependencies: Array<{
    step_id: string;
    step_name: string;
    output: Record<string, any>;
  }>;
}

// ==================== 人工审核相关类型 ====================

/**
 * 审核输出
 */
export interface ReviewOutput {
  summary: string;
  files?: Array<{
    name: string;
    size_bytes: number;
    preview_url?: string;
    download_url?: string;
    id?: string;
  }>;
  reasoning?: string;
  dependencies?: Array<{
    step_id: string;
    step_name: string;
    output_summary: string;
  }>;
}

/**
 * 审核记录（列表项）
 */
export interface ReviewListItem {
  id: string;
  workflow_instance_id: string;
  workflow_name: string;
  step_execution_id: string;
  step_name: string;
  reviewer_id: string;
  reviewer_name: string;
  created_at: string;
  timeout_at: string;
  remaining_time: number;
  review_round: number;
  outputs: ReviewOutput;
}

/**
 * 审核历史
 */
export interface ReviewHistory {
  round: number;
  action: ReviewAction;
  comment: string;
  created_at: string;
  reviewer_name: string;
}

/**
 * 审核记录（详情）
 */
export interface ReviewDetail extends ReviewListItem {
  action?: ReviewAction;
  comment?: string;
  updated_at: string;
  timeout_action?: string;
  history?: ReviewHistory[];
}

/**
 * 审核请求
 */
export interface ReviewRequest {
  comment: string;
}

/**
 * 审核列表查询参数
 */
export interface ReviewListParams extends PaginationParams {
  reviewer_id?: string;
}

/**
 * 审核统计
 */
export interface ReviewStats {
  total_pending: number;
  total_completed_today: number;
  timeout_warnings: number;
  avg_review_time_seconds: number;
  by_action: Record<ReviewAction, number>;
}

// ==================== Agent 管理相关类型 ====================

/**
 * Agent 当前任务
 */
export interface AgentCurrentTask {
  workflow_instance_id: string;
  step_execution_id: string;
  step_name: string;
  progress?: number;
}

/**
 * Agent 统计
 */
export interface AgentStatistics {
  total_tasks: number;
  completed_tasks: number;
  failed_tasks: number;
  avg_task_duration: number;
  success_rate: number;
}

/**
 * Agent（列表项）
 */
export interface AgentListItem {
  id: string;
  name: string;
  display_name: string;
  capabilities: string[];
  status: AgentStatus;
  current_task_id?: string;
  current_workflow_instance_id?: string;
  current_step_execution_id?: string;
  last_heartbeat: string;
  created_at: string;
}

/**
 * Agent（详情）
 */
export interface AgentDetail extends AgentListItem {
  current_task?: AgentCurrentTask;
  config?: Record<string, any>;
  metadata?: Record<string, any>;
  statistics: AgentStatistics;
  updated_at: string;
}

/**
 * Agent 列表查询参数
 */
export interface AgentListParams {
  status?: AgentStatus;
  capability?: string;
  search?: string;
}

/**
 * 批量清理请求
 */
export interface BatchCleanupRequest {
  agent_ids: string[];
}

/**
 * Agent 负载统计
 */
export interface AgentLoadStats {
  agent_id: string;
  agent_name: string;
  status: AgentStatus;
  current_tasks: number;
  running_tasks: number;
  avg_task_duration: number;
}

/**
 * Agent 负载统计响应
 */
export interface AgentLoadStatsResponse {
  data: AgentLoadStats[];
  total_agents: number;
  online_agents: number;
  offline_agents: number;
}

// ==================== 统计与监控相关类型 ====================

/**
 * 工作流统计
 */
export interface WorkflowStats {
  total: number;
  by_status: Record<WorkflowStatus, number>;
  success_rate: number;
  avg_duration_seconds: number;
  by_template: Array<{
    template_id: string;
    template_name: string;
    count: number;
    success_rate: number;
    avg_duration_seconds: number;
  }>;
  time_series: Array<{
    date: string;
    total: number;
    completed: number;
    failed: number;
  }>;
}

/**
 * Agent 统计汇总
 */
export interface AgentStatsSummary {
  total_agents: number;
  online_agents: number;
  offline_agents: number;
  by_agent: Array<{
    agent_id: string;
    agent_name: string;
    total_tasks: number;
    completed_tasks: number;
    failed_tasks: number;
    success_rate: number;
    avg_task_duration: number;
    total_tokens: number;
    estimated_cost_usd: number;
  }>;
  total_tasks: number;
  total_tokens: number;
  total_cost_usd: number;
}

/**
 * 任务统计
 */
export interface TaskStats {
  total: number;
  by_status: Record<string, number>;
  by_category: Record<string, number>;
  avg_duration_seconds: number;
  time_series: Array<{
    date: string;
    total: number;
    completed: number;
    failed: number;
  }>;
}

/**
 * 系统健康检查
 */
export interface HealthCheck {
  status: 'ok' | 'degraded' | 'error';
  timestamp: string;
  components: {
    database: string;
    redis: string;
    gateway: string;
    agents: {
      online: number;
      offline: number;
    };
  };
  version: string;
}

/**
 * 审核统计汇总
 */
export interface ReviewStatsSummary {
  total: number;
  pending: number;
  by_action: Record<ReviewAction, number>;
  avg_review_time_seconds: number;
  timeout_count: number;
  by_reviewer: Array<{
    reviewer_id: string;
    reviewer_name: string;
    total_reviews: number;
    avg_review_time_seconds: number;
  }>;
}

/**
 * 统计查询参数
 */
export interface StatsQueryParams {
  start_date?: string;
  end_date?: string;
  template_id?: string;
}

// ==================== WebSocket 事件类型 ====================

/**
 * WebSocket 订阅消息
 */
export interface WSSubscribeMessage {
  action: 'subscribe' | 'unsubscribe';
  channel: string;
}

/**
 * WebSocket 事件
 */
export interface WSEvent<T = any> {
  event: string;
  channel: string;
  timestamp: string;
  data: T;
}

/**
 * 工作流事件数据
 */
export interface WorkflowEventData {
  workflow_id: string;
  template_name?: string;
  created_by?: string;
  current_step?: CurrentStep;
  duration?: number;
  progress?: number;
}

/**
 * 步骤事件数据
 */
export interface StepEventData {
  workflow_id: string;
  step_id: string;
  step_name?: string;
  agent_id?: string;
  agent_name?: string;
  progress?: number;
  message?: string;
  estimated_remaining?: number;
  duration?: number;
  output?: Record<string, any>;
  review_id?: string;
  reviewer_id?: string;
  timeout_at?: string;
}

/**
 * 审核事件数据
 */
export interface ReviewEventData {
  review_id: string;
  workflow_id: string;
  step_name?: string;
  reviewer_id: string;
  timeout_at?: string;
  remaining_time?: number;
  comment?: string;
}

/**
 * Agent 事件数据
 */
export interface AgentEventData {
  agent_id: string;
  agent_name?: string;
  workflow_id?: string;
  step_id?: string;
  step_name?: string;
}

// ==================== React Flow 节点类型 ====================

/**
 * DAG 节点数据
 */
export interface DAGNodeData {
  label: string;
  status: StepStatus;
  agent: string;
  progress?: number;
  human_review?: boolean;
  estimated_duration?: number;
}

/**
 * DAG 节点
 */
export interface DAGNode {
  id: string;
  type?: string;
  position: { x: number; y: number };
  data: DAGNodeData;
}

/**
 * DAG 边（用于 React Flow）
 */
export interface DAGFlowEdge {
  id: string;
  source: string;
  target: string;
  type?: string;
  animated?: boolean;
}
